### ISSUE TITLE
Question: why use of mutex in models, tensor, and custom op?

### ISSUE THREAD

**Issue by gterzian** [huggingface/candle#960]
trying to familiarize myself with the codebase, surprised by the use of locks in the model caches, for example here https://github.com/huggingface/candle/blob/a36d88325492c5e721b2045b2988cc11ef44fc0e/candle-transformers/src/models/llama.rs#L101

Tried switching to a refcell, it compiled for the llama example. Does it need to be `Sync` for some reason? 

Also, why not make the various `forward` method take a mutable reference? That way you don't need interior mutability.   It seems that the `forward` method are called sequentially(even in the `llama-multiprocess` example, right?). If the model needs to be made `Sync` it can be wrapped in something making it so?

------

I have basically the same question for how [`Storage`](https://github.com/huggingface/candle/blob/a36d88325492c5e721b2045b2988cc11ef44fc0e/candle-core/src/tensor.rs#L39) is used in `_Tensor`. 

----

Lastly, why does `CustomOp1` need to be  `Send + Sync`? I see how it is [called into](https://github.com/huggingface/candle/blob/a36d88325492c5e721b2045b2988cc11ef44fc0e/candle-examples/examples/llama_multiprocess/model.rs#L53) in the `llama-multiprocess` example but again this appears to me as a completely sequential call on the same thread where `TensorParallelRowLinear.forward` is called.

**Comment by EricLBuehler** [comment]
I did not write that code, but perhaps it has a technical CUDA reason? Locking the tensor storagehas some overhead, so unless there is no technical reason I do not think there is any motivation to keep this.

I agree, the `forward` method should probably take a mutable reference. In addition, there is no trainability state exposed, so there is no way to purely evaluate models that have tensors in a `VarBuilder` without gradient accumulation. However, the _Tensor struct has an `is_variable` field - so I can see an API being written for VarBuilder where it goes through each var and changes that state? Just a thought.

**Comment by LaurentMazare** [comment]
One thing to keep in mind is that the cost of the mutex is very negligible compared to the actual computation that takes place so there isn't much to gain by switching to a refcell instead. In the case of `masks` an additional requirement is to share the mask cache across layers hence the ref-counting bit, using a mutex just makes it easy to reason about even if at some point this cache is shared across multiple threads.

It's up to each model to decide how it wants to handle mutable bits, we do have some models that can use mutexes but also models that use a `forward(&mut ...)` method, e.g. [here](https://github.com/huggingface/candle/blob/a36d88325492c5e721b2045b2988cc11ef44fc0e/candle-transformers/src/models/mixformer.rs#L240). There isn't anything magic about the `Module` trait so each model can write a different version of `forward` if it wants to, the default is just that the inner state of the layer is unmodified.

For tensor storage, the cost of locking is also negligible unless you use tensors to store a very small amount of data but that's not what candle is really designed for. We want tensors to be `Send + Sync` so that they can easily be shared across threads, and we use a `RwLock` rather than a `Mutex` as the most common use case is reading and writing is very rare.

`CustomOp1` does not imply `Send + Sync` anymore, the `apply_op1` function requires it because of backprop handling, the backprop graph could be shared across multiple threads by the user. Hower if you don't care about backprop, you could just use [apply_op1_no_bwd](https://github.com/huggingface/candle/blob/a36d88325492c5e721b2045b2988cc11ef44fc0e/candle-core/src/tensor.rs#L2159) and then you don't need `Send + Sync` anymore (but backprop is not supported).

**Comment by gterzian** [comment]
Thanks for the info! 

> the cost of the mutex is very negligible

I'm not thinking about performance at all, rather if there is a mutex I assume there must be some threading otherwise it "looks surprising". 

> We want tensors to be Send + Sync so that they can easily be shared across threads

I can understand that, and one way to do it is a simple `Arc<RwLock<Tensor>>` when it is necessary. Whether it makes sense depends on what tensor operations you want to happen in parallel. If a lot of operations use the internal storage, than locking the tensor as a whole will be similar to just locking the internal storage. If many operations do not use the internal storage and could be done fully in parallel, then you could also put the storage outside of the tensor, and have tensors methods take either a `&mut Storage` or a `&Storage`, which could be achieved by either just having a (mutable) ref to the tensor in a single threaded situation, or otherwise taking a read or write lock, which will deref to the right kind of reference([example](https://github.com/automerge/spanreed/blob/f95edd1f548f3822dcced04607bea794d8887d6e/src/dochandle.rs#L105)). 

The main advantage is that you don't have surprising locks in the code unless those are used(and aren't surprising anymore).

--- 

On a more general point: I'm interested in learning about how these ML training/inference pipeline work. I know little to nothing about ML but I do know about concurrency/distributed, and I also have experience building runtimes that run other people's code in concurrent/distributed context. 

Maybe we can bounce some ideas off of each other? I'd be interested in helping in the development of this framework. 

For example, looking at the `llama-multiprocess` example, I think it would be better to hide all the multiprocessing and CUDA stuff and just have an API where people can call something like `run_multiprocess_pipeline(model, Backend::CUDA)`, or even better `run_parallel_single_node`(so people don't even have to think about threads vs processes). 

Then internally the framework could do all the setup, and finally start running the inference loop while also making certain API's available to the code running inside the loop. For example an API like `all_reduce`, which would take care of all the NCCP stuff for you). 

A bit like a web developer can just write sequential JS/wasm code, and the web engine takes care of a ton of threading/multiprocessing stuff. 

Let me know what you think...

**Comment by Narsil** [comment]
> For example, looking at the llama-multiprocess example, I think it would be better to hide all the multiprocessing and CUDA stuff and just have an API where people can call something like run_multiprocess_pipeline(model, Backend::CUDA), or even better run_parallel_single_node(so people don't even have to think about threads vs processes).

As you say, it's just an example, not a finished product. And what you're describing doesn't belong in candle-core anyway.
The example shows it's easy to add, we have no intent to make it sugary on top.

Also nvidia if you're reading this, please remove the global mutex in CUDA so we can really use multithreading for multi-GPU control instead of having to be multiprocess...

**Comment by ViliamVadocz** [comment]
> Also nvidia if you're reading this, please remove the **global mutex in CUDA** so we can really use multithreading for multi-GPU control instead of having to be multiprocess...

@Narsil Where can I find out more about this global mutex in CUDA?

**Comment by EricLBuehler** [comment]
> It's up to each model to decide how it wants to handle mutable bits, we do have some models that can use mutexes but also models that use a `forward(&mut ...)` method, e.g. [here](https://github.com/huggingface/candle/blob/a36d88325492c5e721b2045b2988cc11ef44fc0e/candle-transformers/src/models/mixformer.rs#L240). There isn't anything magic about the `Module` trait so each model can write a different version of `forward` if it wants to, the default is just that the inner state of the layer is unmodified.

The `Module` trait only allows a `forward(&self)`, so if you want a `forward(&mut self)` you cannot implement `Module`. If `Module` is used for a library that heavily uses generics, then a "Module" with a &mut self forward cannot be a "real" `Module` and therefore cannot be used - a limitation. Of course, interior mutability is possible through `&self` but I feel that `&mut self` would better describe to the caller what is (or may be) happening.

**Comment by Narsil** [comment]
> > Also nvidia if you're reading this, please remove the **global mutex in CUDA** so we can really use multithreading for multi-GPU control instead of having to be multiprocess...
> 
> @Narsil Where can I find out more about this global mutex in CUDA?

https://github.com/coreylowman/cudarc/pull/164#issuecomment-1627647014

**Comment by gterzian** [comment]
> As you say, it's just an example, not a finished product. And what you're describing doesn't belong in candle-core anyway.

Yes I agree, it probably belongs in something akin to `candle-distributed`(like `torch.distributed`). Any plans in that direction? I think there are quite a few things that could be improved if designed from the ground-up(for example a more dynamic leadership concept, as opposed to relying on static ranks, which could improve the story around failure modes and membership changes).

**Comment by LaurentMazare** [comment]
Closing stale issues, feel free to open a fresh one if you're still having problems.
